### PR TITLE
refactor: merge universe setup into world pregen screen

### PR DESCRIPTION
**Pull request by jdrueckert**
### Contains

Merging `UniverseSetupScreen` and `WorldPreGenerationScreen` incl.
* merging `.ui` representations
* merging `.java` classes
* using `UniverseWrapper` for persisting state
* binding UI widgets to persisted state

### How to test

1. Start Terasology
1. Create World via Advanced Game Setup
1. Play around with world setup screen elements

### Outstanding before merging

Based on #5118.

- [X] merge UI elements
- [X] merge widget bindings and screen logic
- [X] use `UniverseWrapper` instead of local copies of target world and seed
- [X] use universe seed and properly randomize seed on re-roll
- [x] fix initial world generator dropdown binding
- [x] remove worlds element in lower right screen section
- [x] update preview on changing seed field content
- [x] memorize modified (manually or via re-roll) seed field content (currently reset always to initial seed)

### Follow-Ups

- [ ] merge WorldSetupScreen into UniverseSetupScreen
  - basically doing #1726 again
- [ ] remove CoreRegistry in favor of Context
- [ ] harden against preview aborts (currently lead to weird state)

### CODE REVIEW

**Comment by jdrueckert** [comment]
@skaldarnar with https://github.com/MovingBlocks/Terasology/pull/5122/commits/bb7fab6bdfb18cb0a63550b7a0ba2436260cf5db I run into the OOM issue again during preview generation:
```
java.util.concurrent.ExecutionException: java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at org.terasology.engine.rendering.nui.layers.mainMenu.preview.FacetLayerPreview.render(FacetLayerPreview.java:130)
	at org.terasology.engine.rendering.nui.layers.mainMenu.UniverseSetupScreen.lambda$updatePreview$7(UniverseSetupScreen.java:416)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.OutOfMemoryError: Java heap space
```

Without it I don't, but I don't get how the changes in that commit would result in running into this kind of a loop there :thinking: any ideas?

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseSetupScreen.java [hunk 1]
```diff
@@ -162,7 +159,7 @@ public WorldGeneratorInfo get() {
                 @Override
                 public void set(WorldGeneratorInfo value) {
                     if (value != null) {
-                        if (universeWrapper.getTargetWorld() == null || !value.getUri().equals(universeWrapper.getTargetWorld().getWorldGenerator().getUri())) {
+                        if (universeWrapper.getWorldConfigurator() == null || !value.getUri().equals(universeWrapper.getWorldGenerator().getUri())) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseSetupScreen.java]: why configurator here?

**Comment by jdrueckert** [comment]
@skaldarnar with this latest commit I'm trying to create the Context already one screen earlier than before, put it into the core registry and consume it from there in UniverseSetupScreen via `@In`. However, the context is empty there, which is why one currently runs into an NPE...
Any idea, why it's not properly picked up from core registry there?

**Comment by skaldarnar** [comment]
I'm not sure that `@In` is working for the `Context` :thinking: or whether that's the right mechanism to use for the context.

The dependency injection with `@In` works if a class requires an instance of a specific other class `T`, and the engine has exactly one instance `t` of that other class `T` it can pass in. If there are multiple or none instances of `T`, the injection cannot work properly.
With the context, I think we explicitly want to have _multiple_ contexts, kinda stacking and dropping them when changing states.

I would try to pass the context explicitly in `setEnvironment` (and rename it to `setContext`) .

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java [hunk 2]
```diff
@@ -779,7 +783,7 @@ private void refreshSelection() {
         updateValidToSelect();
     }
 
-    public void setUniverseWrapper(UniverseWrapper wrapper) {
-        universeWrapper = wrapper;
+    public void setEnvironment(UniverseWrapper wrapper) {
+        CoreRegistry.put(UniverseWrapper.class, wrapper);
```
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java:806]: Theoretically, the idea was to do the following:
```
context.put(UniverseWrapper.class, wrapper);
CoreRegistry.setContext(context);
```
However, this does not work and leads to an NPE in `UniverseSetupScreen.java:182` when attempting to access `UniverseWrapper` from a context that it's not in. At this moment, it's unclear, why this does not work.

**Comment by jdrueckert** [comment]
@skaldarnar from my side this would be ready for review

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseWrapper.java [hunk 3]
```diff
@@ -13,14 +14,15 @@ public class UniverseWrapper {
...
+    private WorldConfigurator worldConfigurator;
+    private WorldGenerator worldGenerator;
 
     public void setSeed(String seed) {
         this.seed = seed;
     }
 
     public String getSeed() {
-        return seed;
+        return seed == null ? "" : seed;
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseWrapper.java]: Is the `null` seed important for anything? If not, we could just initialize it to the empty string as default (and assert that `setSeed` does not accept `null`.

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java [hunk 4]
```diff
@@ -54,82 +51,42 @@ public class WorldSetupScreen extends CoreScreenLayer {
...
      *
      * @param subContext    the new environment created in {@link UniverseSetupScreen}
-     * @param worldSelected the world whose configurations are to be changed.
+     * @param universe      the universe whose world's configurations are to be changed.
      * @throws UnresolvedWorldGeneratorException
      */
-    public void setWorld(Context subContext, WorldSetupWrapper worldSelected)
-            throws UnresolvedWorldGeneratorException {
-        world = worldSelected;
+    public void setWorld(Context subContext, UniverseWrapper universe) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java:83]: I'm annoyed by these HiddenFieldChecks - I'd like to have an overview in how many places this would actually cause trouble right now, and if we find none, disable this check.

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java [hunk 5]
```diff
@@ -54,82 +51,42 @@ public class WorldSetupScreen extends CoreScreenLayer {
...
-        SimpleUri worldGenUri = worldSelected.getWorldGeneratorInfo().getUri();
         environment = context.get(ModuleEnvironment.class);
         context.put(WorldGeneratorPluginLibrary.class, new TempWorldGeneratorPluginLibrary(environment, context));
-        if (world.getWorldGenerator() == null) {
-            worldGenerator = WorldGeneratorManager.createWorldGenerator(worldGenUri, context, environment);
-            world.setWorldGenerator(worldGenerator);
-        } else {
-            worldGenerator = world.getWorldGenerator();
-        }
+        worldGenerator = universe.getWorldGenerator();
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java:88]: The world generator in the universe can no longer be `null`?

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java [hunk 6]
```diff
@@ -117,6 +120,19 @@ public void onOpened() {
         filterModules();
     }
 
+    @Override
+    public void onScreenOpened() {
+        super.onScreenOpened();
+
+        final UIText seed = find("seed", UIText.class);
+        if (seed != null) {
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java]: Hm, if we need to check for `null` here, should we at least log a warning or something in case the `seed` text widget could not be found? I mean, it should be there when the screen is opened, right?

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java [hunk 7]
```diff
@@ -509,12 +525,12 @@ public Boolean get() {
                 getManager().createScreen(MessagePopup.ASSET_URI, MessagePopup.class).
                         setMessage("Error", "Game seed cannot be empty!");
             } else {
-                universeWrapper.setSeed(seed.getText());
+                context.get(UniverseWrapper.class).setSeed(seed.getText());
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java]: We're fetching the universe 3 times from the context here. Can't we do that once and then use that object?

```suggestion
                UniverseWrapper universe = context.get(UniverseWrapper.class);
                universe.setSeed(seed.getText());
```

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java [hunk 8]
```diff
@@ -509,12 +525,12 @@ public Boolean get() {
...
                 saveConfiguration();
-                final GameManifest gameManifest = GameManifestProvider.createGameManifest(universeWrapper, moduleManager, config);
+                final GameManifest gameManifest = GameManifestProvider.createGameManifest(context.get(UniverseWrapper.class), moduleManager, config);
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java]:
```suggestion
                final GameManifest gameManifest = GameManifestProvider.createGameManifest(universe, moduleManager, config);
```

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java [hunk 9]
```diff
@@ -509,12 +525,12 @@ public Boolean get() {
...
                 if (gameManifest != null) {
                     gameEngine.changeState(new StateLoading(gameManifest,
-                            (universeWrapper.getLoadingAsServer()) ? NetworkMode.DEDICATED_SERVER : NetworkMode.NONE));
+                            (context.get(UniverseWrapper.class).getLoadingAsServer()) ? NetworkMode.DEDICATED_SERVER : NetworkMode.NONE));
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java]:
```suggestion
                            (universe.getLoadingAsServer()) ? NetworkMode.DEDICATED_SERVER : NetworkMode.NONE));
```

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java (hunk 2 above)
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java:806]: Do you want to persist this information in the code as comment instead of just here on this PR?

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseSetupScreen.java [hunk 10]
```diff
@@ -215,56 +294,72 @@ private void recursivelyAddModuleDependencies(Set<Name> modules, Name moduleName
...
-        context.put(ModuleManager.class, moduleManager);
-        context.put(UniverseWrapper.class, wrapper);
+    public void setEnvironment() {
+        prepareContext();
 
         DependencyResolver resolver = new DependencyResolver(moduleManager.getRegistry());
         ResolutionResult result = resolver.resolve(config.getDefaultModSelection().listModules());
         if (result.isSuccess()) {
             environment = moduleManager.loadEnvironment(result.getModules(), false);
+            //BlockFamilyLibrary library =  new BlockFamilyLibrary(environment, context);
```
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseSetupScreen.java:337]: Why is this commented out?

**Review by skaldarnar — DISMISSED** [review]

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseSetupScreen.java (hunk 10 above)
Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/UniverseSetupScreen.java:337]: This was originally part of `initAssets` but `library` was never used afterwards.
IIRC I commented it out to see if it breaks anything and wanted to check on possible side-effects it might've been needed for. During my manual tests I didn't see any indicators that commenting it out broke anything.
I can either remove it entirely or replace it with a comment stating that this used to be done in case we see weird behavior in the future that might be related... Which would you prefer?

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java (hunk 4 above)
Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java:83]: I had it run on a iota workspace and it came back with more than 2k findings for hidden fields.
I think most of them are constructors or setters, but it's hard to filter those out.

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java (hunk 5 above)
Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/WorldSetupScreen.java:88]: world generator is set at initialize time (see UniverseSetupScreen#168)
I'd say it not being set is sufficiently unlikely that we don't need the null check anymore.
if you disagree I can re-add it of course

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java (hunk 6 above)
Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java]: tbh I'm not sure, why we need the null check(s) (there's one in AdvancedGameSetupScreen::initialise, too)... I would agree that the text widget should be there when the screen is opened and it should even already be there when it is initialized...
I don't see a valid reason why it should ever not be, so I'll remove the null checks.

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java (hunk 2 above)
Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java:806]: done

File: engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java (hunk 6 above)
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/rendering/nui/layers/mainMenu/advancedGameSetupScreen/AdvancedGameSetupScreen.java]: I think the main reason is that we don't have automated tests that spot NPEs like this. For instance, they can appear if we rename the UI control in the some `.ui` file but not here, or vice versa, move UI controls around but forget about the code, etc.

It will be pretty apparent while testing locally, but we don't have automated tests that "click through all the screens" that would catch that...

**Review by skaldarnar — APPROVED** [review]
